src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,265 @@
+
+        self.num_trains = len(self.dc.train_ids)
+        self.train_ids = self.dc.train_ids
+
+        # check sources
+        self.keys = []
+        for motor_id in product(*coordinates.values()):
+            label = ''.join(f"{n}{v}" for n, v in
+                            zip(coordinates.keys(), motor_id))
+            try:
+                src, key = self.motors[label]
+            except IndexError:
+                raise ValueError(
+                    f"The source for motor {label} is not specified.")
+
+            self.keys.append(self.dc[src, key])
+
+    def positions(self, labelled=True, compressed=False):
+        """Returns the motor positions for all trains.
+
+        Args:
+            labelled (bool):
+                If True, returns the xarray with labelled dimensions,
+                overwise returns numpy.ndarray
+            compressed (bool):
+                If True, returns positions only when they change,
+                overwise positions in all recorded trains
+        Returns:
+            positions (tuple of two numpy.ndarray or xarray.DataArray):
+                The motor positions
+        """
+        train_ids, pos, counts = self._read_positions()
+        if not compressed:
+            pos = np.repeat(pos, counts, axis=0)
+            train_ids = self.train_ids
+
+        if labelled:
+            dims = ["trainId"] + list(self.coordinates.keys())
+            coords = {"trainId": train_ids}
+            coords.update(self.coordinates)
+            return xarray.DataArray(
+                pos, dims=dims, coords=coords, name=self._position_key)
+        else:
+            return train_ids, pos
+
+    def _read_positions(self):
+        """Reads and compresses motor positions."""
+        if hasattr(self, "_positions"):
+            return self._positions
+
+        # read positions
+        pos = np.zeros((self.num_trains, self.num_sources), dtype=float)
+        for source_no, key_data in enumerate(self.keys):
+            pos[:, source_no] = key_data.ndarray()
+        pos = pos.reshape(-1, *self.shape)
+
+        # compress
+        axes = tuple(range(1, pos.ndim))
+        ix = np.flatnonzero(
+            np.insert(np.any(np.diff(pos, axis=0), axis=axes), 0, True))
+
+        # train ids when positions are changed
+        train_ids = np.asarray(self.dc.train_ids)[ix]
+        # count of trains at every position
+        counts = np.diff(np.append(train_ids.astype(int),
+                                   int(self.dc.train_ids[-1]) + 1))
+
+        self._positions = train_ids, pos[ix], counts
+        return self._positions
+
+    def positions_at(self, tid):
+        """Returns motor positions at given train.
+
+        Args:
+            tid (int):
+                Train ID
+
+        Returns:
+            postions (numpy.ndarray):
+                The motor positions
+        """
+        if tid < self.train_ids[0] or self.train_ids[-1] < tid:
```

Comment:
I think this is more correct:
```suggestion
        if tid not in self.train_ids:
```

Otherwise this would still return a value if `tid` happens to be a missing train.

Reply:
True, it is... if we want to accurately, reliably 100% protect.

From other side, it is too expensive (even more expensive than protected code), taking into account that actually should never happen (DAQ populates all trains during aggregation)

If it happens anyway, can we do the same as DAQ  does - populate the missed value with previous? I think this is reasonable compromise, and users/instruments would prefer this instead of dropping image. At least for lit-frame selection, I had to guess of missed data.

I think we can be not so strict here and let users implement more strict protection by themself when it's necessary (or filter data collection properly, e.g. `select` with `require_all=True`)

What do you think?

Reply:
> this actually should never happen (DAQ populates all trains for control data during aggregation)

I believe the DAQ has the possibility to pause recording a run and then resume it later. We can also pool the data from two separate runs (with `.union()`), which will likely have a gap between them. So it's unusual, but definitely not impossible to have missing trains in the middle of control data.

Reply:
I would prefer to be conservative, I don't think our components should ever give data for a train that isn't present. And it seems unlikely that this function will ever be in a hot path so I'd say the performance hit is ok.

Reply:
I would also say that we should never make contiguity assumptions about train IDs for `CONTROL` data, and treat them as arbitrary sets of numbers.

While it is true that it is unlikely to happen for the actual files, it's not uncommon to apply `.select(..., require_all=True)` across all sources early, causing gaps in `CONTROL` data due to gaps in `INSTRUMENT` data, e.g.:

```python
subrun = run.select('*AGIPD*DET*:xtdf', '*MOTOR*')
AGIPD1MQuadrantMotors(subrun)
```

If you're concerned about performance, I wonder if `np.searchsorted` would be faster than `np.isin(unique=True)`. You can use the assumption that train IDs are sorted and unique.

Reply:
In all these cases, there should not be images as well, right?

Reply:
To me that's not a convincing argument :sweat_smile:

Reply:
`searchsorted` wins, if `trains_ids` pre-cast to `ndarray`.

![image](https://github.com/European-XFEL/EXtra/assets/12002097/a4aeaf20-20dd-4354-89b4-5501c209d5e9)

Reply:
Ok, the `tid` protection is strict now. please take a look and let's finish it!